File: backend/app/models/planning.py
"""
Planning Database Models

Models for business planning features including Tasks (to-do list) and
Milestones (company roadmap). Supports multi-tenant isolation with
wallet-based authentication and RAG integration for AI context.
"""

from datetime import datetime
from enum import Enum
from typing import Optional, List
from sqlalchemy import Column, String, Boolean, DateTime, Integer, Text
from sqlalchemy.sql import func
from app.core.database import Base
from pydantic import BaseModel, Field


# Enums for type safety

class TaskPriority(str, Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class TaskCategory(str, Enum):
    SALES = "sales"
    FINANCE = "finance"
    OPERATIONS = "operations"
    MARKETING = "marketing"


class MilestoneStatus(str, Enum):
    PLANNED = "planned"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    AT_RISK = "at_risk"


class TimeframeType(str, Enum):
    QUARTERLY = "quarterly"
    MONTHLY = "monthly"
    CUSTOM = "custom"


class GoalType(str, Enum):
    GROWTH = "growth"
    REVENUE = "revenue"
    PRODUCT = "product"
    OPERATIONS = "operations"


# SQLAlchemy Models

class Task(Base):
    """
    Task (To-Do Item)

    Business task for tracking what needs to be done.
    Supports priorities, categories, and due dates.
    Indexed in Qdrant for AI-powered insights.
    """
    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, index=True)
    wallet_address = Column(String(255), index=True, nullable=False)

    # Task details
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # Categorization
    priority = Column(String(20), default="medium")  # high, medium, low
    category = Column(String(50), default="operations")  # sales, finance, operations, marketing

    # Status
    is_completed = Column(Boolean, default=False)
    completed_at = Column(DateTime(timezone=True), nullable=True)

    # Due date (optional)
    due_date = Column(DateTime(timezone=True), nullable=True)

    # RAG integration tracking
    rag_indexed = Column(Boolean, default=False)
    rag_indexed_at = Column(DateTime(timezone=True), nullable=True)

    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    def __repr__(self):
        return f"<Task(id={self.id}, title='{self.title[:30]}...', priority={self.priority})>"


class Milestone(Base):
    """
    Milestone (Company Roadmap Goal)

    Business goal for tracking company roadmap progress.
    Supports quarterly, monthly, or custom timeframes.
    Indexed in Qdrant for AI-powered insights.
    """
    __tablename__ = "milestones"

    id = Column(Integer, primary_key=True, index=True)
    wallet_address = Column(String(255), index=True, nullable=False)

    # Milestone details
    title = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)

    # Timeline
    timeframe_type = Column(String(20), default="quarterly")  # quarterly, monthly, custom
    timeframe_value = Column(String(100), nullable=False)  # "Q1 2025", "Jan 2025", or custom text
    target_date = Column(DateTime(timezone=True), nullable=True)

    # Progress tracking
    progress_percent = Column(Integer, default=0)  # 0-100
    status = Column(String(20), default="planned")  # planned, in_progress, completed, at_risk

    # Business goal category
    goal_type = Column(String(50), default="growth")  # growth, revenue, product, operations

    # Visual customization
    color = Column(String(7), default="#3B82F6")  # Hex color for visual distinction

    # RAG integration tracking
    rag_indexed = Column(Boolean, default=False)
    rag_indexed_at = Column(DateTime(timezone=True), nullable=True)

    # Timestamps
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    completed_at = Column(DateTime(timezone=True), nullable=True)

    def __repr__(self):
        return f"<Milestone(id={self.id}, title='{self.title[:30]}...', progress={self.progress_percent}%)>"


# Pydantic Schemas for API

class TaskCreate(BaseModel):
    """Schema for creating a new task"""
    wallet_address: str
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    priority: str = "medium"
    category: str = "operations"
    due_date: Optional[datetime] = None


class TaskUpdate(BaseModel):
    """Schema for updating a task"""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    priority: Optional[str] = None
    category: Optional[str] = None
    due_date: Optional[datetime] = None
    is_completed: Optional[bool] = None


class TaskResponse(BaseModel):
    """Schema for task response"""
    id: int
    wallet_address: str
    title: str
    description: Optional[str]
    priority: str
    category: str
    is_completed: bool
    completed_at: Optional[datetime]
    due_date: Optional[datetime]
    created_at: datetime
    updated_at: Optional[datetime]
    # RAG indexing status for visibility
    rag_indexed: Optional[bool] = False
    rag_indexed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class TaskListResponse(BaseModel):
    """Schema for task list response with summary stats"""
    tasks: List[TaskResponse]
    total_count: int
    completed_count: int
    overdue_count: int
    today_count: int


class MilestoneCreate(BaseModel):
    """Schema for creating a new milestone"""
    wallet_address: str
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None
    timeframe_type: str = "quarterly"
    timeframe_value: str = Field(..., min_length=1, max_length=100)
    target_date: Optional[datetime] = None
    goal_type: str = "growth"
    color: Optional[str] = "#3B82F6"


class MilestoneUpdate(BaseModel):
    """Schema for updating a milestone"""
    title: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    timeframe_type: Optional[str] = None
    timeframe_value: Optional[str] = None
    target_date: Optional[datetime] = None
    progress_percent: Optional[int] = Field(None, ge=0, le=100)
    status: Optional[str] = None
    goal_type: Optional[str] = None
    color: Optional[str] = None


class MilestoneResponse(BaseModel):
    """Schema for milestone response"""
    id: int
    wallet_address: str
    title: str
    description: Optional[str]
    timeframe_type: str
    timeframe_value: str
    target_date: Optional[datetime]
    progress_percent: int
    status: str
    goal_type: str
    color: str
    created_at: datetime
    updated_at: Optional[datetime]
    completed_at: Optional[datetime]
    # RAG indexing status for visibility
    rag_indexed: Optional[bool] = False
    rag_indexed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class RoadmapResponse(BaseModel):
    """Schema for roadmap response with grouped milestones"""
    milestones: List[MilestoneResponse]
    current_timeframe: str
    overall_progress: int
    total_count: int
    completed_count: int
    in_progress_count: int
    at_risk_count: int

File: src/types/planning.ts
/**
 * Planning Types - Tasks and Company Roadmap
 *
 * TypeScript interfaces for business planning features.
 * Matches backend models in backend/app/models/planning.py
 */

// =====================================================================
// TASK TYPES
// =====================================================================

export type TaskPriority = 'high' | 'medium' | 'low';

export type TaskCategory = 'sales' | 'finance' | 'operations' | 'marketing';

export interface Task {
  id: number;
  wallet_address: string;
  title: string;
  description?: string;
  priority: TaskPriority;
  category: TaskCategory;
  is_completed: boolean;
  completed_at?: string;
  due_date?: string;
  created_at: string;
  updated_at?: string;
}

export interface TaskCreate {
  wallet_address: string;
  title: string;
  description?: string;
  priority?: TaskPriority;
  category?: TaskCategory;
  due_date?: string;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  priority?: TaskPriority;
  category?: TaskCategory;
  due_date?: string;
  is_completed?: boolean;
}

export interface TaskListResponse {
  tasks: Task[];
  total_count: number;
  completed_count: number;
  overdue_count: number;
  today_count: number;
}

// =====================================================================
// MILESTONE/ROADMAP TYPES
// =====================================================================

export type MilestoneStatus = 'planned' | 'in_progress' | 'completed' | 'at_risk';

export type TimeframeType = 'quarterly' | 'monthly' | 'custom';

export type GoalType = 'growth' | 'revenue' | 'product' | 'operations';

export interface Milestone {
  id: number;
  wallet_address: string;
  title: string;
  description?: string;
  timeframe_type: TimeframeType;
  timeframe_value: string;
  target_date?: string;
  progress_percent: number;
  status: MilestoneStatus;
  goal_type: GoalType;
  color: string;
  created_at: string;
  updated_at?: string;
  completed_at?: string;
}

export interface MilestoneCreate {
  wallet_address: string;
  title: string;
  description?: string;
  timeframe_type?: TimeframeType;
  timeframe_value: string;
  target_date?: string;
  goal_type?: GoalType;
  color?: string;
}

export interface MilestoneUpdate {
  title?: string;
  description?: string;
  timeframe_type?: TimeframeType;
  timeframe_value?: string;
  target_date?: string;
  progress_percent?: number;
  status?: MilestoneStatus;
  goal_type?: GoalType;
  color?: string;
}

export interface RoadmapResponse {
  milestones: Milestone[];
  current_timeframe: string;
  overall_progress: number;
  total_count: number;
  completed_count: number;
  in_progress_count: number;
  at_risk_count: number;
}

// =====================================================================
// UI HELPER TYPES
// =====================================================================

export interface TaskCategoryConfig {
  label: string;
  color: string;
  bgColor: string;
  textColor: string;
}

export interface PriorityConfig {
  label: string;
  color: string;
  dotColor: string;
  bgColor: string;
  textColor: string;
}

export interface MilestoneStatusConfig {
  label: string;
  color: string;
  bgColor: string;
  textColor: string;
}

export interface GoalTypeConfig {
  label: string;
  icon: string;
  color: string;
}

// =====================================================================
// CONFIGURATION OBJECTS
// =====================================================================

export const TASK_CATEGORIES: Record<TaskCategory, TaskCategoryConfig> = {
  sales: {
    label: 'Sales',
    color: 'blue',
    bgColor: 'bg-blue-100',
    textColor: 'text-blue-700',
  },
  finance: {
    label: 'Finance',
    color: 'green',
    bgColor: 'bg-green-100',
    textColor: 'text-green-700',
  },
  operations: {
    label: 'Operations',
    color: 'purple',
    bgColor: 'bg-purple-100',
    textColor: 'text-purple-700',
  },
  marketing: {
    label: 'Marketing',
    color: 'orange',
    bgColor: 'bg-orange-100',
    textColor: 'text-orange-700',
  },
};

export const TASK_PRIORITIES: Record<TaskPriority, PriorityConfig> = {
  high: {
    label: 'Urgent',
    color: 'red',
    dotColor: 'bg-red-500',
    bgColor: 'bg-red-100',
    textColor: 'text-red-700',
  },
  medium: {
    label: 'Medium',
    color: 'amber',
    dotColor: 'bg-amber-500',
    bgColor: 'bg-amber-100',
    textColor: 'text-amber-700',
  },
  low: {
    label: 'Low',
    color: 'gray',
    dotColor: 'bg-gray-400',
    bgColor: 'bg-gray-100',
    textColor: 'text-gray-600',
  },
};

export const MILESTONE_STATUSES: Record<MilestoneStatus, MilestoneStatusConfig> = {
  planned: {
    label: 'Not started',
    color: 'gray',
    bgColor: 'bg-gray-100',
    textColor: 'text-gray-700',
  },
  in_progress: {
    label: 'Working on it',
    color: 'blue',
    bgColor: 'bg-blue-100',
    textColor: 'text-blue-700',
  },
  completed: {
    label: 'Complete',
    color: 'green',
    bgColor: 'bg-green-100',
    textColor: 'text-green-700',
  },
  at_risk: {
    label: 'Needs attention',
    color: 'amber',
    bgColor: 'bg-amber-100',
    textColor: 'text-amber-700',
  },
};

export const GOAL_TYPES: Record<GoalType, GoalTypeConfig> = {
  growth: {
    label: 'Growth',
    icon: 'TrendingUp',
    color: 'blue',
  },
  revenue: {
    label: 'Revenue',
    icon: 'DollarSign',
    color: 'green',
  },
  product: {
    label: 'Product',
    icon: 'Package',
    color: 'purple',
  },
  operations: {
    label: 'Operations',
    icon: 'Settings',
    color: 'orange',
  },
};

// =====================================================================
// UTILITY FUNCTIONS
// =====================================================================

/**
 * Get relative due date string for display
 */
export function getRelativeDueDate(dueDate?: string): string {
  if (!dueDate) return '';

  const due = new Date(dueDate);
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  due.setHours(0, 0, 0, 0);

  const diffDays = Math.floor((due.getTime() - today.getTime()) / (1000 * 60 * 60 * 24));

  if (diffDays < 0) {
    return `Past due by ${Math.abs(diffDays)} day${Math.abs(diffDays) !== 1 ? 's' : ''}`;
  } else if (diffDays === 0) {
    return 'Due today';
  } else if (diffDays === 1) {
    return 'Due tomorrow';
  } else if (diffDays <= 7) {
    return `Due in ${diffDays} days`;
  } else {
    return due.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
  }
}

/**
 * Check if a task is overdue
 */
export function isTaskOverdue(task: Task): boolean {
  if (!task.due_date || task.is_completed) return false;

  const due = new Date(task.due_date);
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  due.setHours(0, 0, 0, 0);

  return due < today;
}

/**
 * Check if a task is due today
 */
export function isTaskDueToday(task: Task): boolean {
  if (!task.due_date || task.is_completed) return false;

  const due = new Date(task.due_date);
  const today = new Date();

  return (
    due.getFullYear() === today.getFullYear() &&
    due.getMonth() === today.getMonth() &&
    due.getDate() === today.getDate()
  );
}

/**
 * Get current quarter string (e.g., "Q1 2025")
 */
export function getCurrentQuarter(): string {
  const now = new Date();
  const quarter = Math.floor(now.getMonth() / 3) + 1;
  return `Q${quarter} ${now.getFullYear()}`;
}

/**
 * Get current month string (e.g., "Jan 2025")
 */
export function getCurrentMonth(): string {
  const now = new Date();
  return now.toLocaleDateString('en-US', { month: 'short', year: 'numeric' });
}
